agent/bls.mjs

```diff
@@ -0,0 +1,2976 @@
+            cc.debug( "Will BLS-sign verified messages." ) + "\n" );
+        let joAccount = optsHandleVerifyAndSign.imaState.chainProperties.sc.joAccount;
+        if( ! joAccount.strURL ) {
+            joAccount = optsHandleVerifyAndSign.imaState.chainProperties.mn.joAccount;
+            if( ! joAccount.strSgxURL )
+                throw new Error( "SGX URL is unknown, cannot sign IMA message(s)" );
+            if( ! joAccount.strBlsKeyName )
+                throw new Error( "BLS keys name is unknown, cannot sign IMA message(s)" );
+        }
+        let rpcCallOpts = null;
+        if( "strPathSslKey" in joAccount &&
+            typeof joAccount.strPathSslKey == "string" &&
+            joAccount.strPathSslKey.length > 0 &&
+            "strPathSslCert" in joAccount &&
+            typeof joAccount.strPathSslCert == "string" &&
+            joAccount.strPathSslCert.length > 0
+        ) {
+            rpcCallOpts = {
+                "cert": fs.readFileSync( joAccount.strPathSslCert, "utf8" ),
+                "key": fs.readFileSync( joAccount.strPathSslKey, "utf8" )
+            };
+        } else {
+            optsHandleVerifyAndSign.details.write(
+                cc.warning( "Will sign via SGX" ) +
+                " " + cc.error( "without SSL options" ) +
+                "\n" );
+        }
+        const signerIndex = optsHandleVerifyAndSign.imaState.nNodeNumber;
+        await rpcCall.create( joAccount.strSgxURL, rpcCallOpts, async function( joCall, err ) {
+            if( err ) {
+                const strErrorMessage =
+                    optsHandleVerifyAndSign.strLogPrefix +
+                    cc.bright( optsHandleVerifyAndSign.strDirection ) + " " +
+                    cc.fatal( "CRITICAL ERROR:" ) +
+                    cc.error( " JSON RPC call to SGX failed, " +
+                        "RPC call was not created, error is: " ) +
+                    cc.warning( owaspUtils.extract_error_message( err ) ) +
+                    "\n";
+                log.write( strErrorMessage );
+                optsHandleVerifyAndSign.details.write( strErrorMessage );
+                if( joCall )
+                    await joCall.disconnect();
+                throw new Error(
+                    "JSON RPC call to SGX failed, RPC call was not created, error is: " +
+                    owaspUtils.extract_error_message( err )
+                );
+            }
+            const joCallSGX = {
+                "jsonrpc": "2.0",
+                "id": randomCallID(),
+                "method": "blsSignMessageHash",
+                "params": {
+                    "keyShareName": joAccount.strBlsKeyName,
+                    "messageHash": optsHandleVerifyAndSign.strMessageHash,
+                    "n": optsHandleVerifyAndSign.nParticipants,
+                    "t": optsHandleVerifyAndSign.nThreshold,
+                    "signerIndex": signerIndex + 0 // 1-based
```

Comment:
Could be replaced by `"signerIndex": signerIndex`

Reply:
This is trick for creating copy of value

Reply:
fun of js))))